urlfcc351a1ba_kuntapavankumar_Z_tgz-pJ8-edu_concordia_dpis_ReplicaTestJ8/src/edu/concordia/dpis/stationserver/FaultyStationServerImpl.java

```diff
@@ -0,0 +1,497 @@
+			} catch (IOException e) {
+				e.printStackTrace();
+				log.error(e.getMessage());
+			}
+			return false;
+		}
+	}
+
+	// A UDPClient to invoke the UDPServer
+	class UDPClient {
+
+		private String ping(String remoteHost, int serverPort)
+				throws FileNotFoundException, IOException {
+			String response = "";
+			Properties properties = new Properties();
+			properties.load(new FileInputStream("udp.properties"));
+			DatagramSocket aSocket = null;
+			try {
+				aSocket = new DatagramSocket();
+				byte[] m = "GET".getBytes();
+				InetAddress aHost = InetAddress.getByName(remoteHost);
+				DatagramPacket request = new DatagramPacket(m, m.length, aHost,
+						serverPort);
+				log.debug(stationType.getStationCode()
+						+ ":UDPClient reuested for number of records");
+				aSocket.send(request);
+				byte[] buffer = new byte[1000];
+				DatagramPacket reply = new DatagramPacket(buffer, buffer.length);
+				aSocket.receive(reply);
+				response = new String(Arrays.copyOfRange(reply.getData(), 0,
+						reply.getLength()));
+				log.debug(stationType.getStationCode() + ":UDPServer replied");
+
+			} catch (SocketException e) {
+				System.out.println("Socket: " + e.getMessage());
+			} catch (IOException e) {
+				System.out.println("IO: " + e.getMessage());
+			} finally {
+				if (aSocket != null) {
+					aSocket.close();
+				}
+			}
+			return response;
+		}
+	}
+
+	// UDPServer
+	class UDPServer {
+		DatagramSocket aSocket = null;
+		final int port;
+
+		public UDPServer(int port) {
+			this.port = port;
+			start();
+		}
+
+		public void start() {
+			new Thread(new Runnable() {
+				@Override
+				public void run() {
+					try {
+						aSocket = new DatagramSocket(port);
+						while (true) {
+							byte[] buffer = new byte[1000];
+							DatagramPacket request = new DatagramPacket(buffer,
+									buffer.length);
+							aSocket.receive(request);
+
+							StringBuffer sb = new StringBuffer();
+							sb.append(stationType.getStationCode());
+							// sb.append(" ");
+							sb.append(countRecords());
+							byte[] payload = sb.toString().getBytes("US-ASCII");
+							DatagramPacket reply = new DatagramPacket(payload,
+									payload.length, request.getAddress(),
+									request.getPort());
+
+							aSocket.send(reply);
+						}
+					} catch (SocketException e) {
+						System.out.println("Socket: " + e.getMessage());
+					} catch (IOException e) {
+						System.out.println("IO: " + e.getMessage());
+					} finally {
+						if (aSocket != null) {
+							aSocket.close();
+						}
+					}
+				}
+			}).start();
+		}
+
+		public void close() {
+			if (aSocket != null) {
+				aSocket.close();
+			}
+		}
+	}
+
+	// TCP SERVER
+	class TCPServer {
+
+		private int port;
+
+		public TCPServer(int port) {
+			this.port = port;
+			start();
+		}
+
+		public void start() {
+			new Thread(new Runnable() {
+
+				@Override
+				public void run() {
+					try {
+						ServerSocket listenSocket = new ServerSocket(port);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *listenSocket*. Execution paths that do not contain closure statements were detected. Ensure *listenSocket* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).